### DISCUSSION TITLE
Teardown mechanism inside a test case

### DISCUSSION THREAD

**Discussion by Exelord** [vitest-dev/vitest#2748] [Ideas]
I often write my tests cases in isolation, using hooks concept. It would be great if we could use a callback like `onCleanup` or `after` inside of a test to register a teardown function.

Example:

```ts
test("checks if file is generates correctly", async () => {
  const path = await mockTmpFolder();
  await generateFile(path)
  // do some assertions
})
```

it would be cool if inside `mockTmpFolder` I could use teardown callback to remove the tmp folder after the test is finished (even if errored).

I am aware I could use `afterEach` but with that I need to use a shared context, eg `this` and then maintain it across my functions.

```ts
describe("mock tmp folder", () => {
  beforeEach(async () => {
    this.path = await mockTmpFolder();
  });

  afterEach(async () => {
    await removeFolder(this.path)
  });

  test("checks if file is generates correctly", async () => {
    await generateFile(this.path)
    // do some assertions
  })
})
```

I assume the test is already aware of a context and it would be possible to use `onCleanup/after` to register functions that would be executed when finished before all other after callbacks. So the usage could look like:

```ts
async function mockTmpFolder() {
  const path = await createDir("tmp/project");
  after(() => removeFolder(path))
  return path;
}
```

I am curios what the community think about it. Maybe such a solution already exists?

**Comment by sheremet-va** [comment]
Well, we do have `onTestFailed` hook, and I think the idea was to extend it, so I don't see any reason not to.

**Reply by Exelord** [reply]
I just noticed that. Awesome!
https://github.com/vitest-dev/vitest/blob/482b72fc5f02da87498a18e0ed844ecb6bc772ca/packages/runner/src/hooks.ts

Would be great to document it as well, as I couldn't find it in API.

Seams like the extension should be easy as well :) I will try to contribute.

**Reply by sheremet-va** [reply]
Should definitely be documented. We have a few features that are not documented for some reason 👀
